Feature: Implement Smart Account and UserOp Validation Logic

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/244

Fixes [ENG-815](https://linear.app/decent-labs/issue/ENG-815/implement-base-smart-account-and-useroperation-validation-contract)

**Summary:**

This PR introduces the `SmartAccountValidationV1` abstract contract, designed to provide robust validation mechanisms for smart accounts (specifically LightAccounts) and their associated UserOperations within an ERC-4337 account abstraction context. It also includes necessary interfaces, mock contracts for thorough testing, and a comprehensive test suite.

**Changes:**

*   **New Abstract Contract `SmartAccountValidationV1.sol`:**
    *   Added `contracts/deployables/account-abstraction/SmartAccountValidationV1.sol`.
    *   This contract provides core logic to:
        *   `validateSmartAccount`: Verify if a given address is a genuine `LightAccount` created by a known `LightAccountFactory`. It checks for contract existence and the ability to call `owner()` on the `LightAccount`, then cross-references the address with the factory.
        *   `validateUserOp`: Validate a `PackedUserOperation`. It first validates the `userOp.sender` using `validateSmartAccount`. Then, it ensures the `userOp.callData` correctly targets the `execute(address,uint256,bytes)` function of the `LightAccount` and extracts the underlying target address and function selector.
    *   Includes custom errors for various invalid states (e.g., `InvalidSmartAccount`, `InvalidUserOpCallDataLength`).
    *   Designed to be initializable, taking a `LightAccountFactory` address.
*   **New Interfaces for Light Accounts:**
    *   `contracts/interfaces/light-account/ILightAccount.sol`: Defines the interface for a `LightAccount`, requiring `owner()` and `execute(...)` functions.
    *   `contracts/interfaces/light-account/ILightAccountFactory.sol`: Defines the interface for a `LightAccountFactory`, requiring a `getAddress(owner, salt)` function.
*   **New Mock Contracts for Testing:**
    *   `contracts/mocks/MockLightAccount.sol`: A mock implementation of `ILightAccount`.
    *   `contracts/mocks/MockInvalidLightAccount.sol`: A mock contract that does *not* correctly implement `ILightAccount` (missing `owner()`), for testing negative paths.
    *   `contracts/mocks/MockLightAccountFactory.sol`: A mock implementation of `ILightAccountFactory` that allows setting expected account addresses for given owners and salts.
    *   `contracts/mocks/MockGaslessTarget.sol`: A simple target contract with dummy functions (`foo`, `bar`) for use in `UserOperation` calldata.
    *   `contracts/mocks/concretes/deployables/account-abstraction/ConcreteSmartAccountValidation.sol`: A concrete implementation of `SmartAccountValidationV1` used to expose internal validation functions for direct testing.
*   **New Test Suite `SmartAccountValidationV1.test.ts`:**
    *   Added `test/deployables/account-abstraction/SmartAccountValidationV1.test.ts`.
    *   Provides extensive test coverage for `SmartAccountValidationV1` via its concrete counterpart:
        *   Validation of legitimate smart accounts.
        *   Rejection of EOAs and contracts not conforming to the `ILightAccount` interface or not created by the expected factory.
        *   Successful validation of well-formed `UserOperation` objects.
        *   Rejection of `UserOperation` objects with invalid senders.
        *   Rejection of `UserOperation` objects with malformed `callData` (too short, incorrect `execute` selector).
        *   Rejection of `UserOperation` objects with malformed inner `callData` (too short).

**Motivation:**

*   To establish a secure and reliable way to verify that a `UserOperation` is initiated by a valid smart contract account (LightAccount) that was deployed through a trusted factory.
*   To ensure that the `callData` within a `UserOperation` is structured as expected, targeting the `execute` function of the LightAccount, thereby preventing malicious or malformed calls from being processed by paymasters or bundlers relying on this validation logic.
*   To provide a reusable and extensible base for smart account validation in systems leveraging ERC-4337.

mudrila commented on test/deployables/account-abstraction/SmartAccountValidationV1.test.ts:
wen `foundry` / `forge`

mudrila commented on test/deployables/account-abstraction/SmartAccountValidationV1.test.ts:
```diff
@@ -0,0 +1,265 @@
+    );
+  });
+
+  describe('validateSmartAccount', function () {
```
I kinda liked more human-friendly descriptions in the previous PR - do you mind adopting these descriptions as well?

adamgall replied:
done

mudrila reviewed (approved):
Man, I bet you've spent at least 3x more time writing tests than implementation 🤣

adamgall committed:
Feature: Implement Smart Account and UserOp Validation Logic

**Summary:**

This PR introduces the `SmartAccountValidationV1` abstract contract, designed to provide robust validation mechanisms for smart accounts (specifically LightAccounts) and their associated UserOperations within an ERC-4337 account abstraction context. It also includes necessary interfaces, mock contracts for thorough testing, and a comprehensive test suite.

**Changes:**

*   **New Abstract Contract `SmartAccountValidationV1.sol`:**
    *   Added `contracts/deployables/account-abstraction/SmartAccountValidationV1.sol`.
    *   This contract provides core logic to:
        *   `validateSmartAccount`: Verify if a given address is a genuine `LightAccount` created by a known `LightAccountFactory`. It checks for contract existence and the ability to call `owner()` on the `LightAccount`, then cross-references the address with the factory.
        *   `validateUserOp`: Validate a `PackedUserOperation`. It first validates the `userOp.sender` using `validateSmartAccount`. Then, it ensures the `userOp.callData` correctly targets the `execute(address,uint256,bytes)` function of the `LightAccount` and extracts the underlying target address and function selector.
    *   Includes custom errors for various invalid states (e.g., `InvalidSmartAccount`, `InvalidUserOpCallDataLength`).
    *   Designed to be initializable, taking a `LightAccountFactory` address.
*   **New Interfaces for Light Accounts:**
    *   `contracts/interfaces/light-account/ILightAccount.sol`: Defines the interface for a `LightAccount`, requiring `owner()` and `execute(...)` functions.
    *   `contracts/interfaces/light-account/ILightAccountFactory.sol`: Defines the interface for a `LightAccountFactory`, requiring a `getAddress(owner, salt)` function.
*   **New Mock Contracts for Testing:**
    *   `contracts/mocks/MockLightAccount.sol`: A mock implementation of `ILightAccount`.
    *   `contracts/mocks/MockInvalidLightAccount.sol`: A mock contract that does *not* correctly implement `ILightAccount` (missing `owner()`), for testing negative paths.
    *   `contracts/mocks/MockLightAccountFactory.sol`: A mock implementation of `ILightAccountFactory` that allows setting expected account addresses for given owners and salts.
    *   `contracts/mocks/MockGaslessTarget.sol`: A simple target contract with dummy functions (`foo`, `bar`) for use in `UserOperation` calldata.
    *   `contracts/mocks/concretes/deployables/account-abstraction/ConcreteSmartAccountValidation.sol`: A concrete implementation of `SmartAccountValidationV1` used to expose internal validation functions for direct testing.
*   **New Test Suite `SmartAccountValidationV1.test.ts`:**
    *   Added `test/deployables/account-abstraction/SmartAccountValidationV1.test.ts`.
    *   Provides extensive test coverage for `SmartAccountValidationV1` via its concrete counterpart:
        *   Validation of legitimate smart accounts.
        *   Rejection of EOAs and contracts not conforming to the `ILightAccount` interface or not created by the expected factory.
        *   Successful validation of well-formed `UserOperation` objects.
        *   Rejection of `UserOperation` objects with invalid senders.
        *   Rejection of `UserOperation` objects with malformed `callData` (too short, incorrect `execute` selector).
        *   Rejection of `UserOperation` objects with malformed inner `callData` (too short).

**Motivation:**

*   To establish a secure and reliable way to verify that a `UserOperation` is initiated by a valid smart contract account (LightAccount) that was deployed through a trusted factory.
*   To ensure that the `callData` within a `UserOperation` is structured as expected, targeting the `execute` function of the LightAccount, thereby preventing malicious or malformed calls from being processed by paymasters or bundlers relying on this validation logic.
*   To provide a reusable and extensible base for smart account validation in systems leveraging ERC-4337.

adamgall committed:
Refactor: Improve Test Descriptions in SmartAccountValidationV1 Tests

**Summary:**

This commit enhances the clarity of test descriptions in the `SmartAccountValidationV1.test.ts` file, making it easier to understand the purpose of each test case.

**Changes:**

* Updated test descriptions to be more descriptive and consistent, including:
  * Changed "valid smart accounts" to "When the smart account is valid".
  * Changed "should validate valid smart accounts" to "should return true and the owner address".
  * Changed "invalid light accounts" to "When the smart account is invalid".
  * Changed "should return false for non-contract addresses" to "should return false and the zero address".
  * Updated other test descriptions for clarity and consistency.

**Motivation:**

* To improve the readability and maintainability of the test suite by providing clearer context for each test case.
